docs/api/sync.md

```diff
@@ -29,6 +32,25 @@ catch(error) {
     console.log(error);
 }
 ```
+### <a name="lock-guard-func-promise-any-promise-any"></a> lock.guard(func: () => Promise\<any>): Promise\<any>
+Run input function and obtain the lock during its execution, returns the promise as the input function returns, or throws error if input function throws. Lock will be released once exception thrown or the promise resolved/rejected.
+```ts
+try {
+    var promise = lock.guard(() => {
```

Comment:
Will the lock be applied during the whole async function call? Is this a meaningful usecase?

Reply:
Will the lock be applied during the whole async function call?
> The answer is YES, if func is an async function, lock.guard() extends the lock until the promise completed.

Is this a meaningful usecase?

> We can think about what is the expected way to do thread synchronization in javascript.
    In other languages, locks are usually applied when accessing exclusive resources, or controlling the order of code flows. For the first scenario, it is very likely to be a async call in javascript.

Reply:
Okay. So let me ask further:

This example will NOT wait for the promise to be resolved:
```
lock.guard(() => {
   readSomethingFromSharedMemory();
   writeAsyncToDisk();
});
```
This example will wait the promise to be resolved:
```
lock.guard(() => {
   readSomethingFromSharedMemory();
   return writeAsyncToDisk();
});
```

Reply:
Yes.

Reply:
I believe the lock class should offer the ability to do synchronization for async calls.

My proposal borrowed some idea from the CallTask and Mocha, which behaved differently based on the return value of the delegate function (whether the return value is a promise or not).

so there are actually 2 problems to discuss:
1. Do we really need to offer the ability for async lock.
2. If we do, what is the best API for that operation.

Reply:
We can go with current proposal. Use return value to decide.